File: backend/src/main/java/com/cvmobile/cv/domain/model/Project.java
package com.cvmobile.cv.domain.model;

import java.time.LocalDate;

/**
 * Projet figurant sur un CV (valeur de domaine immuable).
 *
 * <p>Tous les champs hormis l'identifiant sont facultatifs.
 *
 * @param id           identifiant technique, {@code null} tant que non persiste
 * @param nom          intitule du projet, ou {@code null}
 * @param description  description du projet, ou {@code null}
 * @param technologies technologies employees, ou {@code null}
 * @param lien         lien vers le projet, ou {@code null}
 * @param dateDebut    debut du projet, ou {@code null}
 * @param dateFin      fin du projet, ou {@code null}
 */
public record Project(
        Long id,
        String nom,
        String description,
        String technologies,
        String lien,
        LocalDate dateDebut,
        LocalDate dateFin) {

    /**
     * Construit un projet en normalisant les champs texte.
     */
    public static Project of(
            Long id,
            String nom,
            String description,
            String technologies,
            String lien,
            LocalDate dateDebut,
            LocalDate dateFin) {
        return new Project(
                id,
                DomainText.normalize(nom),
                DomainText.normalize(description),
                DomainText.normalize(technologies),
                DomainText.normalize(lien),
                dateDebut,
                dateFin);
    }

    /**
     * Derive une copie avec une nouvelle description (adaptation IA des
     * variantes), sans muter l'original.
     */
    public Project withDescription(String newDescription) {
        return new Project(
                id, nom, DomainText.normalize(newDescription), technologies, lien,
                dateDebut, dateFin);
    }
}

File: mobile/lib/features/cv/domain/entities/project.dart
import 'copy_with_sentinel.dart';

/// Projet d'un CV, entite de domaine immuable et pure.
final class Project {
  final int? id;
  final String? nom;
  final String? description;
  final String? technologies;
  final String? lien;
  final DateTime? dateDebut;
  final DateTime? dateFin;

  const Project({
    this.id,
    this.nom,
    this.description,
    this.technologies,
    this.lien,
    this.dateDebut,
    this.dateFin,
  });

  Project copyWith({
    Object? id = unsetSentinel,
    Object? nom = unsetSentinel,
    Object? description = unsetSentinel,
    Object? technologies = unsetSentinel,
    Object? lien = unsetSentinel,
    Object? dateDebut = unsetSentinel,
    Object? dateFin = unsetSentinel,
  }) {
    return Project(
      id: identical(id, unsetSentinel) ? this.id : id as int?,
      nom: identical(nom, unsetSentinel) ? this.nom : nom as String?,
      description: identical(description, unsetSentinel)
          ? this.description
          : description as String?,
      technologies: identical(technologies, unsetSentinel)
          ? this.technologies
          : technologies as String?,
      lien: identical(lien, unsetSentinel) ? this.lien : lien as String?,
      dateDebut: identical(dateDebut, unsetSentinel)
          ? this.dateDebut
          : dateDebut as DateTime?,
      dateFin: identical(dateFin, unsetSentinel)
          ? this.dateFin
          : dateFin as DateTime?,
    );
  }

  @override
  bool operator ==(Object other) =>
      other is Project &&
      other.id == id &&
      other.nom == nom &&
      other.description == description &&
      other.technologies == technologies &&
      other.lien == lien &&
      other.dateDebut == dateDebut &&
      other.dateFin == dateFin;

  @override
  int get hashCode => Object.hash(
        id,
        nom,
        description,
        technologies,
        lien,
        dateDebut,
        dateFin,
      );
}
